src/panel/applets/places-indicator/VolumeItem.vala

```diff
@@ -55,28 +72,63 @@ public class VolumeItem : ListItem
                         volume.eject_with_operation.end(res);
                         string safe_remove = _("You can now safely remove");
                         string device_name = volume.get_drive().get_name();
-                        send_message(@"$safe_remove \"$device_name\"", Gtk.MessageType.INFO);
+                        send_message(@"$safe_remove \"$device_name\"");
                     } catch (GLib.Error e) {
-                        send_message(e.message, Gtk.MessageType.ERROR);
+                        send_message(e.message);
                         warning(e.message);
                     }
                 });
             });
         }
 
-        name_button.set_tooltip_text(_("Mount and open \"%s\"").printf(volume.get_name()));
         name_button.clicked.connect(()=> {
-            spin.start();
-            volume.mount.begin(GLib.MountMountFlags.NONE, operation, null, (obj, res)=>{
-                try {
-                    volume.mount.end(res);
-                    open_directory(volume.get_mount().get_root().get_uri());
-                } catch (GLib.Error e) {
-                    send_message(e.message, Gtk.MessageType.ERROR);
-                    warning(e.message);
+            if (unlock_revealer == null) {
+                do_mount(volume);
+            } else {
+                if (!unlock_revealer.get_child_revealed()) {
+                    send_message(_("Enter the encryption passphrase to unlock this volume"));
                 }
-                spin.stop();
-            });
+                unlock_revealer.set_reveal_child(!unlock_revealer.get_child_revealed());
+                operation.get_password_entry().grab_focus();
+            }
         });
     }
+
+    private void do_mount(GLib.Volume volume)
+    {
+        spin.start();
+        volume.mount.begin(GLib.MountMountFlags.NONE, operation, null, (obj, res)=>{
```

Comment:
This really should not be a lambda, the size of it is inappropriate. Should be a real callback that can be debugged